Support using a single key as input for instantiating HDWalletPro…

eggplantzzz opened the pull request:
…vider

eggplantzzz committed:
Support inputting a single key as input for instantiating HDWalletProvider

richardpringle commented:
@gnidan, I'll let you take a look before merging (you might be able to merge yourself too!)

eggplantzzz commented:
I'll ping @gnidan to make sure he gets a chance to look at these changes.

Commit:
Properly pluralize PrivateKey[s]SigningAuthority

Commit:
Increase type explicitness for legacy polymorphism

- Define+use type union `Credentials` for possible legacy options (and
  use `credentials` as the variable name)

- Add type predicate `isPrivateKey` as analog to `isMnemonic` and
  `isPrivateKeys`

- Flatten the cascade of conditionals that translate legacy polymorphism
  into new-style polymorphism.

- Remove type-unsafe check for `privateKey` case inside `isPrivateKeys`

- Stop producing `{ privateKey }` objects during interface translation

Commit:
Remove new privateKey/privateKeys polymorphism

To remove the need to handle this polymorphism after initial constructor
options normalization.

- Permit `{ privateKeys }` instead of `{ privateKey?, privateKeys? }`

- Remove privateKey->privateKeys translation from `getPrivateKeys`

- When determining which interface was used, stop matching `privateKey`
  as valid option.

- Remove test for now-obsolete constructor options interface

gnidan commented on packages/hdwallet-provider/src/constructor/Constructor.ts:
```diff
@@ -16,7 +16,8 @@ export interface MnemonicSigningAuthority {
 export interface PrivateKeySigningAuthority {
-  privateKeys: PrivateKey[];
+  privateKeys?: PrivateKey[];
+  privateKey?: PrivateKey;
```
Note that doing this removes type-safety, since both may be optional and the type-checker would consider it a valid instance of the interface.

In this case it doesn't matter, since it the code that generates these objects isn't capable of generating an object with neither of those properties, but still... I went with this approach in the first place with the goal to be as explicit as possible ;)

gnidan commented on packages/hdwallet-provider/src/constructor/getOptions.ts:
```diff
@@ -2,22 +2,26 @@ import { MnemonicPhrase, PrivateKey } from "./types";
 
 // check that the first argument is a mnemonic phrase
 const isMnemonicPhrase = (
   mnemonicPhraseOrPrivateKeys: MnemonicPhrase | PrivateKey[]
```
This type is misleading if it doesn't mention `PrivateKey` as a third option... seems like there's room to come up with a name for this.

I propose `credentials: Credentials`.

gnidan commented on packages/hdwallet-provider/src/constructor/getOptions.ts:
```diff
@@ -2,22 +2,26 @@ import { MnemonicPhrase, PrivateKey } from "./types";
 ): mnemonicPhraseOrPrivateKeys is PrivateKey[] =>
-  mnemonicPhraseOrPrivateKeys instanceof Array;
+  mnemonicPhraseOrPrivateKeys instanceof Array ||
+  typeof mnemonicPhraseOrPrivateKeys === "string";
```
This is going to break type safety!!!

This function tells TypeScript, "I will return a boolean value, and, based on that boolean, you will whether the type you asked about is a `PrivateKey[]`"... by returning `true` for strings, you break that promise, and TS is left none-the-wiser.

gnidan commented on packages/hdwallet-provider/src/constructor/getOptions.ts:
```diff
@@ -2,22 +2,26 @@ import { MnemonicPhrase, PrivateKey } from "./types";
+// check that the first argument is a list of private keys or a single key
 const isPrivateKeys = (
-  mnemonicPhraseOrPrivateKeys: MnemonicPhrase | PrivateKey[]
+  mnemonicPhraseOrPrivateKeys: PrivateKey | PrivateKey[]
```
This should stay the same as the argument to `isMnemonicPhrase` (well, both should have all 3 options, really)... the idea of these functions are to take something that can be one of many things and tell the type-checker how it knows when it's figured out which one it is.

gnidan commented on packages/hdwallet-provider/src/constructor/getOptions.ts:
```diff
@@ -26,12 +30,13 @@ const getSigningAuthorityOptions = (
   } else if (isPrivateKeys(mnemonicPhraseOrPrivateKeys)) {
-    return {
-      privateKeys: mnemonicPhraseOrPrivateKeys
-    };
+    return typeof mnemonicPhraseOrPrivateKeys === "string" ?
+      { privateKeys: [ mnemonicPhraseOrPrivateKeys ] } :
+      { privateKeys: mnemonicPhraseOrPrivateKeys };
```
Instead of this, it's more clear to have a third conditional for `isPrivateKey`.

Going with my suggestion, to define a `LegacyConstructor.Credentials` type:
```typescript
export type Credentials = MnemonicPhrase | PrivateKey | PrivateKey[];
```

Then we'd define three separate type predicates (see [docs about this](https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates)):
- `isMnemonic`
- `isPrivateKey`
- `isPrivateKeys`

The idea is that each of these takes a value of any of the three allowed types and returns a boolean based on which one it is. This gets used both at runtime, for conditional logic, and at compile-time, to tell TS not to barf on `for (const privateKey of mnemonicPhraseOrPrivateKeys)` if you wrap it in `if (isPrivateKeys(mnemonicPhraseOrPrivateKeys)) { ... }`.

So then, in this function, we could just lean on having type predicates for each possibility, and move the ternary operator into a new `else if` that's at the same top-level.

gnidan commented on packages/hdwallet-provider/src/index.ts:
```diff
@@ -117,6 +117,13 @@ class HDWalletProvider {
       ethUtilValidation(privateKeys);
     } // no need to handle else case here, since matchesNewOptions() covers it
 
+    if (this.addresses.length === 0) {
+      throw new Error(
+        `Could not create addresses from your mnemonic or private key(s). ` +
+          `Please check that your inputs are correct.`
+      );
+    }
```
This is a good check to include! (What prompted you to add it? I'm curious.)

gnidan commented on packages/hdwallet-provider/test/provider.test.ts:
```diff
@@ -25,84 +25,204 @@ describe("HD Wallet Provider", function () {
     });
 
-    web3.eth.getBlockNumber((err, number) => {
+    it("provides for a private key", async () => {
+      const privateKey =
+        "3f841bf589fdf83a521e55d51afddc34fa65351161eead24f064855fc29c9580"; //random valid private key generated with ethkey
+      provider = new HDWalletProvider({
+        privateKey,
+        providerOrUrl: `http://localhost:${port}`
+      });
+      web3.setProvider(provider);
+
+      const addresses = provider.getAddresses();
+      assert.equal(addresses[0], "0xc515db5834d8f110eee96c3036854dbf1d87de2b");
+      addresses.forEach(address => {
+        assert(EthUtil.isValidAddress(address), "invalid address");
+      });
+
+      const number = await web3.eth.getBlockNumber();
```
I think we should disallow the single-`privateKey` case in the new interface, while we still can.

My reasoning:

- It's very little cost for users to do `{ privateKeys: [privateKey] }` instead of `{ privateKey }`.
  - Naturally, `privateKeys` works perfectly well for both cases, so there's no negative impact except possible confusion.
  - We can remove risk of confusion by specifying "how to pass a single private key" in the README (although I'm not sure it's necessary).
- Including this option adds extra logic:
  - Something has to distinguish this case at run-time (like you've added to `getPrivateKeys`).
  - The type descriptions need to account for the extra possibility, and that gets into risky areas because TS forces you to do either `{ a } | { b }` or `{ a?, b? }`. The latter here is type unsafe; the former adds something else to deal with.

eggplantzzz replied:
I'm not opposed to not supporting `privateKey`. It seems kind of unnecessary as it is just a special case of `PrivateKey[]`.  But I always hear your voice in the back of my head saying "we should handle/accept as wide a variety of input as possible" haha.

gnidan commented on packages/hdwallet-provider/test/provider.test.ts:
```diff
@@ -25,84 +25,204 @@ describe("HD Wallet Provider", function () {
-      "3f841bf589fdf83a521e55d51afddc34fa65351161eead24f064855fc29c9580",
-      "9549f39decea7b7504e15572b2c6a72766df0281cea22bd1a3bc87166b1ca290"
-    ];
+  describe("instantiating with non-positional arguments", () => {
```
Thanks for making tests for this!

gnidan reviewed (changes requested):
@eggplantzzz got some revisions for you. I opened a PR [here](https://github.com/trufflesuite/truffle/pull/3340) because I figured it'd be easier to show what I mean by example.

Sorry for being unclear in my write-up of this remaining change to @richardpringle's PR. I appreciate your offering to take this change off my plate... I would've missed a few of the changes here and probably let this sit another month! 😮 🙇

gnidan committed:
Merge pull request #3340 from trufflesuite/suggest/support-single-key

Use more explicit code style for new HDWalletProvider constructor interface

gnidan commented on packages/hdwallet-provider/src/constructor/getOptions.ts:
```diff
@@ -2,37 +2,51 @@ import { MnemonicPhrase, PrivateKey } from "./types";
+  validateMnemonic(credentials, wordlist);
 
-// check that the first argument is a list of private keys
+// check that the first argument is a list of private keys or a single key
```
Oops comment here is wrong, I'll fix
```suggestion
// check that the first argument is a list of private keys
```

gnidan commented on packages/hdwallet-provider/src/constructor/getOptions.ts:
```diff
@@ -2,37 +2,51 @@ import { MnemonicPhrase, PrivateKey } from "./types";
+): credentials is PrivateKey[] =>
+  credentials instanceof Array;
+
+const isPrivateKey = (
```
```suggestion
// check that the first argument is a single private key (default case for invalid mnemonics)
const isPrivateKey = (
```

gnidan reviewed (changes requested):
(needs some comment fixing)

gnidan committed:
Fix code comments for legacy polymorphism

- Make sure `isPrivateKeys`' comment doesn't lie
- Add comment for `isPrivateKey`

gnidan reviewed (approved):
I think this is the last of it before the original PR is a-go!

Will wait for the tests to turn green then hit merge on this one
